### ISSUE TITLE
DATETIME retrieved via FreeTDS ODBC includes "bonus" (incorrect) fractional seconds precision

### ISSUE THREAD

**Issue by gordthompson** [mkleehammer/pyodbc#448]
Xubuntu 18.04, 64-bit
Python version 3.6.5, 64-bit
unixODBC 2.3.7
pyodbc 4.0.24
msodbcsql17 17.02.0000
FreeTDS ODBC v1.00.94

test table:
CREATE TABLE dt_test (id INT PRIMARY KEY, dt DATETIME)

One row of data

```
id  dt
--  -----------------------
 1  2018-08-22 15:45:32.123
```

Test code:

```python
result = crsr.execute("SELECT dt FROM dt_test WHERE id=1").fetchval()
print(repr(result))
```

When run with an "ODBC Driver 17 for SQL Server" connection, the test code prints the correct result

```
datetime.datetime(2018, 8, 22, 15, 45, 32, 123000)
```

OTOH, when run with a FreeTDS ODBC connection, it prints a result with extra microseconds

```
datetime.datetime(2018, 8, 22, 15, 45, 32, 123333)
```

Interestingly, if I create a "return the raw bytes" output converter function ...

```python
def handle_datetime(dt_value):
    return dt_value
```

... and declare it with ...

```python
cnxn.add_output_converter(pyodbc.SQL_TYPE_TIMESTAMP, handle_datetime)
```

... then in both cases it prints the same thing:

```
b'D\xa9\x00\x00\xf5\xb2\x03\x01'
```

ODBC trace logs:
[msodbcsql17.log](https://github.com/mkleehammer/pyodbc/files/2312280/msodbcsql17.log)
[freetds.log](https://github.com/mkleehammer/pyodbc/files/2312281/freetds.log)

**Comment by v-chojas** [comment]
According to [the TDS spec](https://msdn.microsoft.com/en-us/library/ee780895.aspx) the datetime contains "One 4-byte unsigned integer that represents the _number of one three-hundredths of a second (300 counts per second)_ elapsed since 12 AM that day", so 123333 is actually the more correct value if you continue the fractional expansion, but it only has a precision of (slightly less than) 3.

There were a few breaking changes in SQL Server 2016 that are related to this: http://support.microsoft.com/help/4010261

**Comment by gordthompson** [comment]
Understood. So does each ODBC driver do the math to convert the wire protocol bytes into a standard value that represents a SQL_TYPE_TIMESTAMP?

I discovered the issue because [this test](https://github.com/mkleehammer/pyodbc/blob/5634d6b19c8b1c3d71b6d18901013697549b9b0a/tests2/freetdstests.py#L529) was failing using the latest stable FreeTDS ODBC driver, and the way the test was written suggests that once upon a time FreeTDS ODBC did limit its values to millisecond precision.

Also, do all SQL Server types that map to SQL_TYPE_TIMESTAMP have the same precision and rounding behaviour? In other words, should the driver do the math like my output converter function [here](https://stackoverflow.com/a/48351121/2144390)? If so, then I probably should open a FreeTDS issue.

**Comment by v-chojas** [comment]
> So does each ODBC driver do the math to convert the wire protocol bytes into a standard value that represents a SQL_TYPE_TIMESTAMP?

SQL_TYPE_TIMESTAMP is a SQL / server type, not a C type. If you mean the C types SQL_C_TIMESTAMP/SQL_C_TYPE_TIMESTAMP, then yes --- the driver converts from TDS format into the given structure type.

> Also, do all SQL Server types that map to SQL_TYPE_TIMESTAMP have the same precision and rounding behaviour?

No, the obvious counterexample being datetime2(n) and smalldatetime. The precision and scale can vary and are returned in IPD/IRD fields.


See the links below for more information.

https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-date-time/data-type-support-for-odbc-date-and-time-improvements

https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-date-time/metadata-parameter-and-result

**Comment by gordthompson** [comment]
Okay, thanks. Seeing as how ODBC seems to be aware of how many "Decimal Digits" the seconds should have ...
 
```
[ODBC][13496][1535041175.535408][SQLDescribeColW.c][467]
        Exit:[SQL_SUCCESS]                
            Column Name = [dt](unicode)                
            Data Type = 0x7fff85d90392 -> 93                
            Column Size = 0x7fff85d90398 -> 23                
            Decimal Digits = 0x7fff85d90394 -> 3                
            Nullable = 0x7fff85d90396 -> 1
...
[ODBC][13496][1535041175.535525][SQLGetData.c][237]
        Entry:
            Statement = 0x13c5e20
            Column Number = 1
            Target Type = 93 SQL_TYPE_TIMESTAMP
            Buffer Length = 16
            Target Value = 0x7fff85d90590
            StrLen Or Ind = 0x7fff85d90588
```

... is it reasonable to expect FreeTDS ODBC to do the rounding?

**Comment by gordthompson** [comment]
By the way, I found the following structure [here](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/c-data-types)

```
struct tagTIMESTAMP_STRUCT {  
   SQLSMALLINT year;  
   SQLUSMALLINT month;  
   SQLUSMALLINT day;  
   SQLUSMALLINT hour;  
   SQLUSMALLINT minute;  
   SQLUSMALLINT second;  
   SQLUINTEGER fraction;
} TIMESTAMP_STRUCT;
```

So it's the driver's job to unpack whatever TDS delivers into that structure, correct?

**Comment by v-chojas** [comment]
> ... is it reasonable to expect FreeTDS ODBC to do the rounding?

It can be argued either way, because the actual fractional value stored in the datetime is 37/300 and it has a precision of 3, so anything beyond the 3rd decimal place is theoretically meaningless. You could say it should be 0.123000 because that's what you get if you zero-extend 0.123, or that it should be 0.123333 because that's the actual fraction stored. The msodbcsql17 chose one way, FreeTDS chose another. They're both acceptable because from the standpoint of precision, the digits beyond the 3rd _don't actually matter_ (and the 3rd can only be 0, 3, or 7.)

> So it's the driver's job to unpack whatever TDS delivers into that structure, correct?

Yes.

**Comment by gordthompson** [comment]
Understood. I chose a better example when I reported the issue to FreeTDS

```python
s = '1900-01-01 00:00:00.007'
```

because the msodbcsql and FreeTDS fractional values are more obviously different (.007000 vs .006666).

And even in my original case the extra digits do matter because

```python
datetime(2018, 8, 22, 15, 45, 32, 123000) != datetime(2018, 8, 22, 15, 45, 32, 123333)
```

which is why the freetdstests were failing.

IMO what really needs to happen is for FreeTDS to calculate the `fraction` value in `TIMESTAMP_STRUCT` to conform with the [datetime documentation](https://docs.microsoft.com/en-us/sql/t-sql/data-types/datetime-transact-sql), i.e., not

2 / 300.0 * 1000000000 = 6666666 (or maybe 6666667)

but

round(2 / 300.0, 3) * 1000000000 = 7000000
